<a href="https://colab.research.google.com/github/UnaRam/pythonFundamemtals/blob/main/Una_15_1_Encoding_and_Dummy_Coding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Encoding Data  
When we encode data, we remodel a numeric column in a dataframe to be on a standard scale (0 or 1).   
For example if we had a column of BMI scores, we could encode that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

For example:  

` def encode_bmi(df):`       
> `if df['bmi'] >= 25:`  
> >  `return 1`   

>`else`:  
> >`return 0`  

`df["bmi"] = df.apply(encode_bmi, axis=1)`    

### Challenge 1 - prepare dataset for encoding 
---
1. Read Covid vaccination data from the `by_country` sheet in the Excel file at this link : https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true
2. Find out which columns have missing values
3. Remove all rows with missing data in the total_vaccination column  
4. Remove all rows with missing data in the daily_vaccinations_per_million 
4. find the median daily_vaccinations_per_million, storing this in a variable for use later     


**Test output**:  
1. dataframe is saved in a variable
2. 
```
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
    Column                               Non-Null Count  Dtype         
                                
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_fully_vaccinated_per_hundred  6158 non-null   float64       
 11  daily_vaccinations_per_million       14796 non-null  float64       
 12  vaccines                             14994 non-null  object        
 13  source_name                          14994 non-null  object        
 14  source_website                       14994 non-null  object        
dtypes: datetime64[ns](1), float64(9), object(5)
memory usage: 1.7+ MB
```
3. 9011 rows × 15 columns  
4. 8815 rows * 15 columns 
5. 6.65 



**Read Covid vaccination data** from the by_country sheet in the Excel file at this link : https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true

should display a dataframe
check rows x columns


In [ ]:
import pandas as pd
covid_url = "https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"

# Task 1 
def read_file_df(url, sheet_to_read):
  df = pd.read_excel(url,sheet_name=sheet_to_read)
  return df

raw_covid_df = read_file_df(covid_url,"by_country")

print(raw_covid_df.shape)
print(raw_covid_df.info())
#print(raw_covid_df.describe())
display(raw_covid_df)



(14994, 15)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10

country iso_code       date  total_vaccinations  people_vaccinated  \
0      Afghanistan      AFG 2021-02-22                 0.0                0.0   
1      Afghanistan      AFG 2021-02-23                 NaN                NaN   
2      Afghanistan      AFG 2021-02-24                 NaN                NaN   
3      Afghanistan      AFG 2021-02-25                 NaN                NaN   
4      Afghanistan      AFG 2021-02-26                 NaN                NaN   
...            ...      ...        ...                 ...                ...   
14989     Zimbabwe      ZWE 2021-04-28            458013.0           388021.0   
14990     Zimbabwe      ZWE 2021-04-29            477597.0           400771.0   
14991     Zimbabwe      ZWE 2021-04-30            500342.0           414735.0   
14992     Zimbabwe      ZWE 2021-05-01            520299.0           428135.0   
14993     Zimbabwe      ZWE 2021-05-02            524199.0           430068.0   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                          NaN                     NaN                 NaN   
1                          NaN                     NaN              1367.0   
2                          NaN                     NaN              1367.0   
3                          NaN                     NaN              1367.0   
4                          NaN                     NaN              1367.0   
...                        ...                     ...                 ...   
14989                  69992.0                 24074.0             17860.0   
14990                  76826.0                 19584.0             17971.0   
14991                  85607.0                 22745.0             19194.0   
14992                  92164.0                 19957.0             21171.0   
14993                  94131.0                  3900.0             19169.0   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                                0.00                           0.00   
1                                 NaN                            NaN   
2                                 NaN                            NaN   
3                                 NaN                            NaN   
4                                 NaN                            NaN   
...                               ...                            ...   
14989                            3.08                           2.61   
14990                            3.21                           2.70   
14991                            3.37                           2.79   
14992                            3.50                           2.88   
14993                            3.53                           2.89   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                      NaN                             NaN   
1                                      NaN                            35.0   
2                                      NaN                            35.0   
3                                      NaN                            35.0   
4                                      NaN                            35.0   
...                                    ...                             ...   
14989                                 0.47                          1202.0   
14990                                 0.52                          1209.0   
14991                                 0.58                          1291.0   
14992                                 0.62                          1424.0   
14993                                 0.63                          1290.0   

                 vaccines                source_name  \
0      Oxford/AstraZeneca  Government of Afghanistan   
1      Oxford/AstraZeneca  Government of Afghanistan   
2      Oxford/AstraZeneca  Government of Afghanistan   
3      Oxford/AstraZeneca  Government of Afghanistan   
4      Oxford/Astra

**Find out which columns have missing values**
separate functions 

a)to check a df for rows with missing values in a particular column

b) test this 

In [ ]:
# Task 2
def find_col_missing_values(df,col_to_check):
  
  return df[col_to_check].isnull().values.any()
  


def test_missing(col_to_check, missing):
  if missing == True:
    print(col_to_check, "has missing rows\n")
  else:
    print(col_to_check, "no columns found with missing rows\n\n")

#test - check if any missing
col = "daily_vaccinations_per_million"
test_missing(col, find_col_missing_values(raw_covid_df,col))

col= "total_vaccinations" 
test_missing(col, find_col_missing_values(raw_covid_df,col))

rows, cols = raw_covid_df.shape
print("Raw rows ",rows,"\n\n")
print(raw_covid_df.info())





daily_vaccinations_per_million has missing rows

total_vaccinations has missing rows

Raw rows  14994 


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null 

Remove all rows with missing data in the total_vaccination column
Remove all rows with missing data in the daily_vaccinations_per_million

combine the above two tasks as will be using the same function
 need to filter out rows for a specific columns

check that the number of rows reduces 

In [ ]:
# filter out null rows for a specific column
def filter_out_nulls(df, col_to_filter):
  print("col ", col_to_filter)
  new_df = df[~df[col_to_filter].isnull()]
  return new_df

# first filter for total_vaccinations
filtered_covid_df = filter_out_nulls(raw_covid_df,'total_vaccinations')
print(filtered_covid_df.info())
rows, cols = filtered_covid_df.shape
print("Filtered rows ",rows)

# then for daily_vaccinations_per_million
filtered_covid_df = filter_out_nulls(filtered_covid_df,'daily_vaccinations_per_million')  #when should I create a new df???
print(filtered_covid_df.info())
rows, cols = filtered_covid_df.shape
print("Filtered rows ",rows)

col  total_vaccinations
<class 'pandas.core.frame.DataFrame'>
Int64Index: 9011 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              9011 non-null   object        
 1   iso_code                             9011 non-null   object        
 2   date                                 9011 non-null   datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8294 non-null   float64       
 5   people_fully_vaccinated              6157 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   8816 non-null   float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8294 non-null   float64

find the median daily_vaccinations_per_million, storing this in a variable for use later

In [ ]:
def find_df_median(df,col):
  median = df[col].median()
  return median

median_daily_v_per_mill =find_df_median(filtered_covid_df,'daily_vaccinations_per_million')
print(median_daily_v_per_mill)



1915.5


**different from expected - has the question changed?? - this is the median for total_vaccinations per hundred**

In [ ]:
filtered_covid_df['daily_vaccinations_per_million'].describe()
filtered_covid_df.describe()

total_vaccinations  people_vaccinated  people_fully_vaccinated  \
count        8.816000e+03       8.126000e+03             6.143000e+03   
mean         5.073578e+06       3.255572e+06             1.583705e+06   
std          2.054483e+07       1.197302e+07             6.848034e+06   
min          0.000000e+00       0.000000e+00             1.000000e+00   
25%          6.587150e+04       5.604150e+04             2.373450e+04   
50%          4.550545e+05       3.492675e+05             1.747800e+05   
75%          1.996336e+06       1.412589e+06             6.897980e+05   
max          2.753380e+08       1.470470e+08             1.047747e+08   

       daily_vaccinations_raw  daily_vaccinations  \
count            7.575000e+03        8.816000e+03   
mean             1.347569e+05        1.145023e+05   
std              5.211912e+05        4.414027e+05   
min              0.000000e+00        0.000000e+00   
25%              2.888500e+03        2.693500e+03   
50%              1.558300e+04        1.367750e+04   
75%              6.249850e+04        5.236525e+04   
max              1.160100e+07        7.205286e+06   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
count                     8816.000000                    8126.000000   
mean                        15.783548                      11.432801   
std                         23.385756                      15.392827   
min                          0.000000                       0.000000   
25%                          1.480000                       1.360000   
50%                          6.650000                       5.050000   
75%                         20.155000                      14.890000   
max                        211.080000                     111.320000   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  
count                          6143.000000                     8816.000000  
mean                              5.667215                     3352.155626  
std                              10.062239                     4980.979866  
min                               0.000000                        0.000000  
25%                               0.630000                      658.000000  
50%                               2.350000                     1915.500000  
75%                               6.315000                     4388.000000  
max                              99.760000                   118759.000000

### Challenge 2 - encode daily vaccinations 
---

Write a function to encode daily vaccinations per million, where values greater than or equal to median = 1 and values less than median = 0 

**Test output**: 

using describe()
```
count    8816.000000
mean        0.991493
std         0.091847
min         0.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         1.000000
Name: daily_vaccinations_per_million, dtype: float64
```

**is this expected value correct? if using median shouldn't it then be centred around that median??**

perform encoding
the functions takes a dictionary list of keywords
in this case use median and key(or col?)

In [ ]:
# Task
def encode_daily(df, **kwds):
  median = kwds['median']
  column = kwds['col']
  if df[column] >= median:
    return 1
  else:
    return 0


filtered_covid_df["daily_vacs_mill_enc"]=filtered_covid_df.apply(encode_daily,axis=1 , col = "daily_vaccinations_per_million", median = median_daily_v_per_mill)
print("daily vacs per mill enc:\t",filtered_covid_df['daily_vacs_mill_enc'].describe())
print("orig column",filtered_covid_df['daily_vaccinations_per_million'].describe())
# display(filtered_covid_df)


daily vacs per mill enc:	 count    8816.000000
mean        0.500000
std         0.500028
min         0.000000
25%         0.000000
50%         0.500000
75%         1.000000
max         1.000000
Name: daily_vacs_mill_enc, dtype: float64
orig column count      8816.000000
mean       3352.155626
std        4980.979866
min           0.000000
25%         658.000000
50%        1915.500000
75%        4388.000000
max      118759.000000
Name: daily_vaccinations_per_million, dtype: float64


### Challenge 3 - Encoding total vaccinations   
---
The United Kingdom has been praised for its fast vaccine rollout. 
1. Find the minimum total vaccinations for the United Kingdom 
2. Write a function to encode total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1 
3. Display the unique countries for which total vaccinated is at the same rate or more than the UK

**Test output**:

1. 1402432.0
2. `df['total_vaccinations'].describe()` should output:
```
count    9011.00000
mean        0.29808
std         0.45744
min         0.00000
25%         0.00000
50%         0.00000
75%         1.00000
max         1.00000
Name: total_vaccinations, dtype: float64
```
3. 
```
array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)
```




Find the minimum total vaccinations for the United Kingdom

create a function to do this as it is often used

use a separate filter as this helps my brain cope with the syntax

use the unfiltered df raw_covid_df

In [ ]:
# Task 1
# Find the minimum total vaccinations for the United Kingdom
def find_min_val_col(df,col):
  min_val = df[col].min()
  return(min_val)

#filter forvalue in a column and return new df
def filter_for_val_in_col(df,col,val):
  filter = (df[col]==val)
  new_df = df[filter]
  return new_df

uk_df = filter_for_val_in_col(raw_covid_df,"country", "United Kingdom")
uk_df['total_vaccinations'].describe()
uk_min_tot_vacc = find_min_val_col(uk_df,"total_vaccinations")
print("UK Min total vaccinations",uk_min_tot_vacc)



UK Min total vaccinations 1402432.0


Write a function to encode total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1

start with the raw_covid_df and re-filter for total_vaccinations

2. `df['total_vaccinations'].describe()` should output:


count    9011.00000

mean        0.29808

std         0.45744

min         0.00000

25%         0.00000

50%         0.00000

75%         1.00000

max         1.00000

Name: total_vaccinations, dtype: float64

In [ ]:

# Task 2
def encode_total_vacc(df, **kwds):
  minimum = kwds['min']
  column = kwds['col']
  if df[column] >= minimum:
    return 1
  else:
    return 0


# first filter for total_vaccinations using previously created function


temp_df =filter_out_nulls(raw_covid_df,'total_vaccinations')
total_vacc_df = temp_df.copy()  #otherwise get a warning about trying to set a value on a slice from a df 

total_vacc_df['total_vacc_gte_uk']=total_vacc_df.apply(encode_total_vacc,axis=1 , col = "total_vaccinations", min = uk_min_tot_vacc)
print(" unique countries for which total vaccinated is at the same rate or more than the UK")
print("total_vacc_gte_uk:\t",total_vacc_df['total_vacc_gte_uk'].describe())
# print("orig column",total_vacc_df['total_vaccinations'].describe())




col  total_vaccinations
 unique countries for which total vaccinated is at the same rate or more than the UK
total_vacc_gte_uk:	 count    9011.00000
mean        0.29808
std         0.45744
min         0.00000
25%         0.00000
50%         0.00000
75%         1.00000
max         1.00000
Name: total_vacc_gte_uk, dtype: float64


Display the unique countries for which total vaccinated is at the same rate or more than the UK

 ***Wouldn't it be better to look at this normalised by population?***

Apply a filter to see if the encoded total_vacc is 1


expected values
```
array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)
```



In [ ]:
# Task 3
filter = total_vacc_df['total_vacc_gte_uk']==1
gte_uk_df = total_vacc_df[filter]
gte_uk_df['country'].describe
countries_gte_uk = gte_uk_df['country'].unique()
print(countries_gte_uk)
print(len(countries_gte_uk))
# print(gte_uk_df['country']['total_vacc_gte_uk'])

['Argentina' 'Australia' 'Austria' 'Azerbaijan' 'Bangladesh' 'Belgium'
 'Brazil' 'Cambodia' 'Canada' 'Chile' 'China' 'Colombia' 'Czechia'
 'Denmark' 'Dominican Republic' 'England' 'Finland' 'France' 'Germany'
 'Greece' 'Hong Kong' 'Hungary' 'India' 'Indonesia' 'Ireland' 'Israel'
 'Italy' 'Japan' 'Kazakhstan' 'Malaysia' 'Mexico' 'Morocco' 'Nepal'
 'Netherlands' 'Norway' 'Pakistan' 'Peru' 'Philippines' 'Poland'
 'Portugal' 'Qatar' 'Romania' 'Russia' 'Saudi Arabia' 'Scotland' 'Serbia'
 'Singapore' 'Slovakia' 'South Korea' 'Spain' 'Sweden' 'Switzerland'
 'Thailand' 'Turkey' 'United Arab Emirates' 'United Kingdom'
 'United States' 'Uruguay' 'Wales']
59


are the uk nations also included separately?

get list of all included countries


In [ ]:

all_countries = raw_covid_df['country'].unique()
print(len(all_countries), all_countries)


195 ['Afghanistan' 'Albania' 'Algeria' 'Andorra' 'Angola' 'Anguilla'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba' 'Australia' 'Austria'
 'Azerbaijan' 'Bahamas' 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus'
 'Belgium' 'Belize' 'Bermuda' 'Bhutan' 'Bolivia' 'Bosnia and Herzegovina'
 'Botswana' 'Brazil' 'Brunei' 'Bulgaria' 'Cambodia' 'Cameroon' 'Canada'
 'Cape Verde' 'Cayman Islands' 'Chile' 'China' 'Colombia' 'Congo'
 'Costa Rica' "Cote d'Ivoire" 'Croatia' 'Curacao' 'Cyprus' 'Czechia'
 'Democratic Republic of Congo' 'Denmark' 'Djibouti' 'Dominica'
 'Dominican Republic' 'Ecuador' 'Egypt' 'El Salvador' 'England'
 'Equatorial Guinea' 'Estonia' 'Eswatini' 'Ethiopia' 'Faeroe Islands'
 'Falkland Islands' 'Fiji' 'Finland' 'France' 'Gabon' 'Gambia' 'Georgia'
 'Germany' 'Ghana' 'Gibraltar' 'Greece' 'Greenland' 'Grenada' 'Guatemala'
 'Guernsey' 'Guinea' 'Guyana' 'Honduras' 'Hong Kong' 'Hungary' 'Iceland'
 'India' 'Indonesia' 'Iran' 'Iraq' 'Ireland' 'Isle of Man' 'Israel'
 'Italy' 'Jamaica' 'J

### Challenge 4 - create new series of total vaccinations for each manufacturer
---

To create a new column in your dataframe:

`df['new_column'] = ...`

For example:

* to duplicate an existing column
  * `df['new_column'] = df['old_column']`
* to add two columns together 
  * `df['new_column'] = df['column1'] + df['column2']`
* to make a percentages column 
  * `df['new_column'] = (df['column1']/df['column1].sum()) * 100`

  
1. read data from 'by_manufacturer' sheet from Covid data 
2. find the sum of total vaccinations for each manufacturer
3. create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations 
4. find the median percentage 
5. create a new column called 'encoded_percentages' which duplicates the percentages column
6. encode the encoded_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0 


**Test output**:

1.
2. 
```
vaccine
Johnson&Johnson        264839828
Moderna               5548036383
Oxford/AstraZeneca     539433203
Pfizer/BioNTech       8690461304
Sinovac                604660293
Name: total_vaccinations, dtype: int64
```
3. 
```
	location	date	vaccine	total_vaccinations	percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055
...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423
3296 rows × 5 columns
```
4. 0.0011110194374896931
5. 
```
location	date	vaccine	total_vaccinations	percentage_of_total	encoded_percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055
...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423
3296 rows × 6 columns
```
6. 
```
	location	date	vaccine	total_vaccinations	percentages	encode	encoded
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003	0
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033	0
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053	0
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055	0
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055	0
...	...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095	1
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504	1
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519	1
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413	1
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423	1
3296 rows × 7 columns
```



read data from 'by_manufacturer' sheet from Covid data

re-use existing function

test by inspection of resulting dataframe - contains manufacturer info


In [ ]:
covid_url = "https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"

# Task 1 
def read_file_df(url, sheet_to_read):
  df = pd.read_excel(url,sheet_name=sheet_to_read)
  return df

raw_manufact_df = read_file_df(covid_url,"by_manufacturer")

print(raw_manufact_df.shape)
print(raw_manufact_df.info())

display(raw_manufact_df)

(3296, 4)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3296 entries, 0 to 3295
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   location            3296 non-null   object        
 1   date                3296 non-null   datetime64[ns]
 2   vaccine             3296 non-null   object        
 3   total_vaccinations  3296 non-null   int64         
dtypes: datetime64[ns](1), int64(1), object(2)
memory usage: 103.1+ KB
None


location       date          vaccine  total_vaccinations
0             Chile 2020-12-24  Pfizer/BioNTech                 420
1             Chile 2020-12-25  Pfizer/BioNTech                5198
2             Chile 2020-12-26  Pfizer/BioNTech                8338
3             Chile 2020-12-27  Pfizer/BioNTech                8649
4             Chile 2020-12-28  Pfizer/BioNTech                8649
...             ...        ...              ...                 ...
3291  United States 2021-05-01          Moderna           105947940
3292  United States 2021-05-01  Pfizer/BioNTech           129013657
3293  United States 2021-05-02  Johnson&Johnson             8374395
3294  United States 2021-05-02          Moderna           106780082
3295  United States 2021-05-02  Pfizer/BioNTech           130252779

[3296 rows x 4 columns]

find the sum of total vaccinations for each manufacturer

expected = 
    vaccine

    Johnson&Johnson        264839828

    Moderna               5548036383
    
    Oxford/AstraZeneca     539433203
    
    Pfizer/BioNTech       8690461304
    
    Sinovac                604660293
  
Name: total_vaccinations, dtype: int64

.sum() sums all numeric columns - only one in this case


In [ ]:
sums_per_man_df = raw_manufact_df.groupby('vaccine').sum()
print(sums_per_man_df)
overall_total_vacc =sums_per_man_df['total_vaccinations'].sum()
print("overall vaccinations: ",overall_total_vacc)

                    total_vaccinations
vaccine                               
Johnson&Johnson              264839828
Moderna                     5548036383
Oxford/AstraZeneca           539433203
Pfizer/BioNTech             8690461304
Sinovac                      604660293
overall vaccinations:  15647431011


create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations


In [ ]:
manuf_df = raw_manufact_df.copy()
manuf_df['pct_vacc']=100*manuf_df['total_vaccinations']/overall_total_vacc
display(manuf_df)
print(manuf_df.describe())

location       date          vaccine  total_vaccinations  pct_vacc
0             Chile 2020-12-24  Pfizer/BioNTech                 420  0.000003
1             Chile 2020-12-25  Pfizer/BioNTech                5198  0.000033
2             Chile 2020-12-26  Pfizer/BioNTech                8338  0.000053
3             Chile 2020-12-27  Pfizer/BioNTech                8649  0.000055
4             Chile 2020-12-28  Pfizer/BioNTech                8649  0.000055
...             ...        ...              ...                 ...       ...
3291  United States 2021-05-01          Moderna           105947940  0.677095
3292  United States 2021-05-01  Pfizer/BioNTech           129013657  0.824504
3293  United States 2021-05-02  Johnson&Johnson             8374395  0.053519
3294  United States 2021-05-02          Moderna           106780082  0.682413
3295  United States 2021-05-02  Pfizer/BioNTech           130252779  0.832423

[3296 rows x 5 columns]

       total_vaccinations     pct_vacc
count        3.296000e+03  3296.000000
mean         4.747400e+06     0.030340
std          1.578003e+07     0.100847
min          0.000000e+00     0.000000
25%          1.571925e+04     0.000100
50%          1.738460e+05     0.001111
75%          1.790371e+06     0.011442
max          1.302528e+08     0.832423


find the median percentage


expected: 0.0011110194374896931


In [ ]:
median_pct_vacc=manuf_df['pct_vacc'].median()
print(median_pct_vacc)

0.0011110194374896931


create a new column called 'encoded_percentages' which duplicates the percentages column

In [ ]:
manuf_df["encoded_percentages"]=manuf_df["pct_vacc"]
display(manuf_df)


location       date          vaccine  total_vaccinations  pct_vacc  \
0             Chile 2020-12-24  Pfizer/BioNTech                 420  0.000003   
1             Chile 2020-12-25  Pfizer/BioNTech                5198  0.000033   
2             Chile 2020-12-26  Pfizer/BioNTech                8338  0.000053   
3             Chile 2020-12-27  Pfizer/BioNTech                8649  0.000055   
4             Chile 2020-12-28  Pfizer/BioNTech                8649  0.000055   
...             ...        ...              ...                 ...       ...   
3291  United States 2021-05-01          Moderna           105947940  0.677095   
3292  United States 2021-05-01  Pfizer/BioNTech           129013657  0.824504   
3293  United States 2021-05-02  Johnson&Johnson             8374395  0.053519   
3294  United States 2021-05-02          Moderna           106780082  0.682413   
3295  United States 2021-05-02  Pfizer/BioNTech           130252779  0.832423   

      encoded_percentages  
0                0.000003  
1                0.000033  
2                0.000053  
3                0.000055  
4                0.000055  
...                   ...  
3291             0.677095  
3292             0.824504  
3293             0.053519  
3294             0.682413  
3295             0.832423  

[3296 rows x 6 columns]

encode the encoded_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0

In [ ]:
def encode_pcts(df, **kwds):
  median = kwds['med']
  column = kwds['col']
  if df[column] >= median:
    return 1
  else:
    return 0


# first filter for total_vaccinations using previously created function


# temp_df =filter_out_nulls(raw_covid_df,'total_vaccinations')
# total_vacc_df = temp_df.copy()  #otherwise get a warning about trying to set a value on a slice from a df 

manuf_df['encoded_percentages']=manuf_df.apply(encode_pcts,axis=1 , col = "pct_vacc", med = median_pct_vacc)
display(manuf_df)

location       date          vaccine  total_vaccinations  pct_vacc  \
0             Chile 2020-12-24  Pfizer/BioNTech                 420  0.000003   
1             Chile 2020-12-25  Pfizer/BioNTech                5198  0.000033   
2             Chile 2020-12-26  Pfizer/BioNTech                8338  0.000053   
3             Chile 2020-12-27  Pfizer/BioNTech                8649  0.000055   
4             Chile 2020-12-28  Pfizer/BioNTech                8649  0.000055   
...             ...        ...              ...                 ...       ...   
3291  United States 2021-05-01          Moderna           105947940  0.677095   
3292  United States 2021-05-01  Pfizer/BioNTech           129013657  0.824504   
3293  United States 2021-05-02  Johnson&Johnson             8374395  0.053519   
3294  United States 2021-05-02          Moderna           106780082  0.682413   
3295  United States 2021-05-02  Pfizer/BioNTech           130252779  0.832423   

      encoded_percentages  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
...                   ...  
3291                    1  
3292                    1  
3293                    1  
3294                    1  
3295                    1  

[3296 rows x 6 columns]

### Exercise 8 - encode daily vaccinations 

1. find the median daily vaccinations per 1 million 
2. write a function to encode daily vaccinations per 1 million, where values greater than or equal to median = 1 and values less than median = 0 

Output: 

1. 1915.5
2. 
```
0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: daily_vaccinations_per_million, Length: 9011, dtype: int64
```

In [ ]:
# # find the median daily vaccinations per 1 million
# daily_vacc_per_mill_df = filter_out_nulls(raw_covid_df,'daily_vaccinations_per_million')  #when should I create a new df???
# median_daily_vacc_per_mill = daily_vacc_per_mill_df['daily_vaccinations_per_million'].median()
# print(median_daily_vacc_per_mill)
# print(daily_vacc_per_mill_df['daily_vaccinations_per_million'].describe())

col  daily_vaccinations_per_million
1475.0
count     14796.000000
mean       2972.163625
std        4720.846894
min           0.000000
25%         347.000000
50%        1475.000000
75%        3908.250000
max      118759.000000
Name: daily_vaccinations_per_million, dtype: float64


In [ ]:
def encode_daily_mill(df, **kwds):
  median = kwds['med']
  column = kwds['col']
  if df[column] >= median:
    return 1
  else:
    return 0


# first filter for daily_vaccinations_per_million using previously created function
temp_df = filter_out_nulls(raw_covid_df,'daily_vaccinations_per_million')  #when should I create a new df???
daily_vacc_per_mill_df = temp_df.copy()

# find the median daily vaccinations per 1 million
median_daily_vacc_per_mill = daily_vacc_per_mill_df['daily_vaccinations_per_million'].median()
print(median_daily_vacc_per_mill)
# print(daily_vacc_per_mill_df['daily_vaccinations_per_million'].describe())

#create new column with encoded
daily_vacc_per_mill_df['encoded_daily_per_million']=daily_vacc_per_mill_df.apply(encode_pcts,axis=1 , col = "daily_vaccinations_per_million", med = median_daily_vacc_per_mill)
display(daily_vacc_per_mill_df)

col  daily_vaccinations_per_million
1475.0


country iso_code       date  total_vaccinations  people_vaccinated  \
1      Afghanistan      AFG 2021-02-23                 NaN                NaN   
2      Afghanistan      AFG 2021-02-24                 NaN                NaN   
3      Afghanistan      AFG 2021-02-25                 NaN                NaN   
4      Afghanistan      AFG 2021-02-26                 NaN                NaN   
5      Afghanistan      AFG 2021-02-27                 NaN                NaN   
...            ...      ...        ...                 ...                ...   
14989     Zimbabwe      ZWE 2021-04-28            458013.0           388021.0   
14990     Zimbabwe      ZWE 2021-04-29            477597.0           400771.0   
14991     Zimbabwe      ZWE 2021-04-30            500342.0           414735.0   
14992     Zimbabwe      ZWE 2021-05-01            520299.0           428135.0   
14993     Zimbabwe      ZWE 2021-05-02            524199.0           430068.0   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
1                          NaN                     NaN              1367.0   
2                          NaN                     NaN              1367.0   
3                          NaN                     NaN              1367.0   
4                          NaN                     NaN              1367.0   
5                          NaN                     NaN              1367.0   
...                        ...                     ...                 ...   
14989                  69992.0                 24074.0             17860.0   
14990                  76826.0                 19584.0             17971.0   
14991                  85607.0                 22745.0             19194.0   
14992                  92164.0                 19957.0             21171.0   
14993                  94131.0                  3900.0             19169.0   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
1                                 NaN                            NaN   
2                                 NaN                            NaN   
3                                 NaN                            NaN   
4                                 NaN                            NaN   
5                                 NaN                            NaN   
...                               ...                            ...   
14989                            3.08                           2.61   
14990                            3.21                           2.70   
14991                            3.37                           2.79   
14992                            3.50                           2.88   
14993                            3.53                           2.89   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
1                                      NaN                            35.0   
2                                      NaN                            35.0   
3                                      NaN                            35.0   
4                                      NaN                            35.0   
5                                      NaN                            35.0   
...                                    ...                             ...   
14989                                 0.47                          1202.0   
14990                                 0.52                          1209.0   
14991                                 0.58                          1291.0   
14992                                 0.62                          1424.0   
14993                                 0.63                          1290.0   

                 vaccines                source_name  \
1      Oxford/AstraZeneca  Government of Afghanistan   
2      Oxford/AstraZeneca  Government of Afghanistan   
3      Oxford/AstraZeneca  Government of Afghanistan   
4      Oxford/AstraZeneca  Government of Afghanistan   
5      Oxford/Astra

In [ ]:
filter = daily_vacc_per_mill_df['encoded_daily_per_million']==1
gte_uk_mill_df = daily_vacc_per_mill_df[filter]
countries_gte_uk_mill = gte_uk_mill_df['country'].unique()
print(len(countries_gte_uk_mill), countries_gte_uk_mill)

125 ['Albania' 'Andorra' 'Anguilla' 'Antigua and Barbuda' 'Argentina' 'Aruba'
 'Australia' 'Austria' 'Azerbaijan' 'Bahamas' 'Bahrain' 'Barbados'
 'Belarus' 'Belgium' 'Belize' 'Bermuda' 'Bhutan' 'Bolivia' 'Brazil'
 'Bulgaria' 'Cambodia' 'Canada' 'Cayman Islands' 'Chile' 'China'
 'Colombia' 'Costa Rica' 'Croatia' 'Curacao' 'Cyprus' 'Czechia' 'Denmark'
 'Dominica' 'Dominican Republic' 'Ecuador' 'El Salvador' 'England'
 'Equatorial Guinea' 'Estonia' 'Eswatini' 'Faeroe Islands'
 'Falkland Islands' 'Fiji' 'Finland' 'France' 'Germany' 'Gibraltar'
 'Greece' 'Greenland' 'Grenada' 'Guernsey' 'Guyana' 'Hong Kong' 'Hungary'
 'Iceland' 'India' 'Indonesia' 'Ireland' 'Isle of Man' 'Israel' 'Italy'
 'Jamaica' 'Jersey' 'Jordan' 'Kazakhstan' 'Kuwait' 'Latvia' 'Lebanon'
 'Liechtenstein' 'Lithuania' 'Luxembourg' 'Macao' 'Maldives' 'Malta'
 'Mauritius' 'Mexico' 'Monaco' 'Mongolia' 'Montenegro' 'Montserrat'
 'Morocco' 'Nauru' 'Nepal' 'Netherlands' 'New Zealand' 'Northern Cyprus'
 'Northern Ireland' 'Norway'

so 125 countries have the same total_vaccines per million people as UK

### Exercise 9 - Encoding vaccinations per hundred  
---
The United Kingdom has been praised for its fast vaccine rollout. 
1. find the minimum total vaccinations for the United Kingdom 
2. save this value in a variable rounded down to an integer
3. write a function to encode total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1 
4. display the countries which total vaccinated is at the same rate or more than the UK

Output:

1. 1402432.0
2. 1402432
3. `df['people_vaccinated_per_hundred']` should output:
```
0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: total_vaccinations, Length: 9011, dtype: int64
```
4. 
```
array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)
```




find the minimum total vaccinations for the United Kingdom
save this value in a variable rounded down to an integer

In [ ]:
def encode_total_hundred(df, **kwds):
  minimum = kwds['min']
  column = kwds['col']
  if df[column] >= median:
    return 1
  else:
    return 0


# first filter for daily_vaccinations_per_million using previously created function
temp_df = filter_out_nulls(raw_covid_df,'total_vaccinations_per_hundred')  #when should I create a new df???
total_vacc_per_hundred_df = temp_df.copy()
#apply filter for just uk
uk_total_per_hundred_df = filter_for_val_in_col(total_vacc_per_hundred_df,"country", "United Kingdom")
print("uk_total_per_hundred_df")
display(uk_total_per_hundred_df)

# find the median daily vaccinations per 1 million
min_total_vacc_per_hundred = uk_total_per_hundred_df['total_vaccinations_per_hundred'].min()
print("min total vacc per hundred uk",min_total_vacc_per_hundred)


#create new column with encoded
total_vacc_per_hundred_df['people_vaccinated_per_hundred']=total_vacc_per_hundred_df.apply(encode_total_hundred,axis=1 , col = "total_vaccinations_per_hundred", min = min_total_vacc_per_hundred)
print("total_vacc_per_hundred_df")
display(total_vacc_per_hundred_df)

col  total_vaccinations_per_hundred
uk_total_per_hundred_df


country iso_code       date  total_vaccinations  \
14329  United Kingdom      GBR 2021-01-03           1402432.0   
14336  United Kingdom      GBR 2021-01-10           2677971.0   
14337  United Kingdom      GBR 2021-01-11           2843815.0   
14338  United Kingdom      GBR 2021-01-12           3067541.0   
14339  United Kingdom      GBR 2021-01-13           3356229.0   
...               ...      ...        ...                 ...   
14443  United Kingdom      GBR 2021-04-27          47540984.0   
14444  United Kingdom      GBR 2021-04-28          48138009.0   
14445  United Kingdom      GBR 2021-04-29          48752144.0   
14446  United Kingdom      GBR 2021-04-30          49319518.0   
14447  United Kingdom      GBR 2021-05-01          49834997.0   

       people_vaccinated  people_fully_vaccinated  daily_vaccinations_raw  \
14329          1380430.0                  22002.0                     NaN   
14336          2286572.0                 391399.0                     NaN   
14337          2431648.0                 412167.0                165844.0   
14338          2639309.0                 428232.0                223726.0   
14339          2918252.0                 437977.0                288688.0   
...                  ...                      ...                     ...   
14443         33959908.0               13581076.0                495593.0   
14444         34094048.0               14043961.0                597025.0   
14445         34216616.0               14535528.0                614135.0   
14446         34362205.0               14957313.0                567374.0   
14447         34505380.0               15329617.0                515479.0   

       daily_vaccinations  total_vaccinations_per_hundred  \
14329                 NaN                            2.07   
14336            182220.0                            3.94   
14337            179880.0                            4.19   
14338            185810.0                            4.52   
14339            201020.0                            4.94   
...                   ...                             ...   
14443            517918.0                           70.03   
14444            526822.0                           70.91   
14445            534262.0                           71.81   
14446            529487.0                           72.65   
14447            503641.0                           73.41   

       people_vaccinated_per_hundred  people_fully_vaccinated_per_hundred  \
14329                           2.03                                 0.03   
14336                           3.37                                 0.58   
14337                           3.58                                 0.61   
14338                           3.89                                 0.63   
14339                           4.30                                 0.65   
...                              ...                                  ...   
14443                          50.02                                20.01   
14444                          50.22                                20.69   
14445                          50.40                                21.41   
14446                          50.62                                22.03   
14447                          50.83                                22.58   

       daily_vaccinations_per_million  \
14329                             NaN   
14336                          2684.0   
14337                          2650.0   
14338                          2737.0   
14339                          2961.0   
...                               ...   
14443                          7629.0   
14444                          7760.0   
14445                          7870.0   
14446                          7800.0   
14447                          7419.0   

                                           vaccines  \
14329  Moderna, Oxford/AstraZeneca, Pfizer/BioNTech   
14336  Moderna, Oxford/AstraZeneca, Pfizer/BioN

min total vacc per hundred uk 2.07
total_vacc_per_hundred_df


country iso_code       date  total_vaccinations  people_vaccinated  \
0      Afghanistan      AFG 2021-02-22                 0.0                0.0   
6      Afghanistan      AFG 2021-02-28              8200.0             8200.0   
22     Afghanistan      AFG 2021-03-16             54000.0            54000.0   
44     Afghanistan      AFG 2021-04-07            120000.0           120000.0   
59     Afghanistan      AFG 2021-04-22            240000.0           240000.0   
...            ...      ...        ...                 ...                ...   
14989     Zimbabwe      ZWE 2021-04-28            458013.0           388021.0   
14990     Zimbabwe      ZWE 2021-04-29            477597.0           400771.0   
14991     Zimbabwe      ZWE 2021-04-30            500342.0           414735.0   
14992     Zimbabwe      ZWE 2021-05-01            520299.0           428135.0   
14993     Zimbabwe      ZWE 2021-05-02            524199.0           430068.0   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                          NaN                     NaN                 NaN   
6                          NaN                     NaN              1367.0   
22                         NaN                     NaN              2862.0   
44                         NaN                     NaN              3000.0   
59                         NaN                     NaN              8000.0   
...                        ...                     ...                 ...   
14989                  69992.0                 24074.0             17860.0   
14990                  76826.0                 19584.0             17971.0   
14991                  85607.0                 22745.0             19194.0   
14992                  92164.0                 19957.0             21171.0   
14993                  94131.0                  3900.0             19169.0   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                                0.00                              0   
6                                0.02                              0   
22                               0.14                              0   
44                               0.31                              0   
59                               0.62                              0   
...                               ...                            ...   
14989                            3.08                              0   
14990                            3.21                              0   
14991                            3.37                              0   
14992                            3.50                              0   
14993                            3.53                              0   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                      NaN                             NaN   
6                                      NaN                            35.0   
22                                     NaN                            74.0   
44                                     NaN                            77.0   
59                                     NaN                           206.0   
...                                    ...                             ...   
14989                                 0.47                          1202.0   
14990                                 0.52                          1209.0   
14991                                 0.58                          1291.0   
14992                                 0.62                          1424.0   
14993                                 0.63                          1290.0   

                 vaccines                source_name  \
0      Oxford/AstraZeneca  Government of Afghanistan   
6      Oxford/AstraZeneca  Government of Afghanistan   
22     Oxford/AstraZeneca  Government of Afghanistan   
44     Oxford/AstraZeneca  Government of Afghanistan   
59     Oxford/Astra

### Exercise 10 - create new series of total vaccinations percentages
---

To create a new column in your dataframe:

`df['new_column'] = ...`

For example:

* to duplicate an existing column
  * `df['new_column'] = df['old_column']`
* to add two columns together 
  * `df['new_column'] = df['column1'] + df['column2']`
* to make a percentages column 
  * `df['new_column'] = (df['column1']/df['column1].sum()) * 100`  
  


1. read data from 'by_manufacturer' sheet from Covid data 
2. find the sum of total vaccinations for each manufacturer
3. create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations 
4. find the median percentage 
5. create a new column called 'encoded_percentages' which duplicates the percentages column
6. encode the encoded_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0 


Output:

1.
2. 
```
vaccine
Johnson&Johnson        264839828
Moderna               5548036383
Oxford/AstraZeneca     539433203
Pfizer/BioNTech       8690461304
Sinovac                604660293
Name: total_vaccinations, dtype: int64
```
3. 
```
	location	date	vaccine	total_vaccinations	percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055
...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423
3296 rows × 5 columns
```
4. 0.0011110194374896931
5. 
6. 
```
	location	date	vaccine	total_vaccinations	percentages	encode	encoded
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003	0
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033	0
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053	0
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055	0
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055	0
...	...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095	1
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504	1
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519	1
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413	1
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423	1
3296 rows × 7 columns
```



# Reflection
----

## What skills have you demonstrated in completing this notebook?

This worksheet was an introduction to encoding (I would think of it as setting a flag....)
very useful way of dealing with comparing data

***new skills***
* passing keywords into a function
* refresher on using the built-in fuctions for descriptive stats from a df


## What caused you the most difficulty?

Worksheet was mostly straighforward.

As usual, I need to take care when reading the question